src/shadowbox/server/server_access_key.ts

```diff
@@ -90,73 +123,162 @@ export class ServerAccessKeyRepository implements AccessKeyRepository {
+    if (!quota || quota.quotaBytes < 1 || quota.windowHours < 1) {
+      return false;
+    }
+    const accessKey = this.getAccessKey(id);
+    if (!accessKey) {
+      return false;
+    }
+    accessKey.quota = quota;
+    try {
+      this.saveAccessKeys();
+      const quotaStausChanged = await this.updateAccessKeyQuotaStaus(accessKey);
+      if (quotaStausChanged) {
+        // Reflect the access key quota status if it changed with the new quota.
+        await this.updateServer();
+      }
+    } catch (error) {
+      return false;
+    }
+    return true;
+  }
+
+  async removeAccessKeyQuota(id: AccessKeyId): Promise<boolean> {
+    const accessKey = this.getAccessKey(id);
+    if (!accessKey) {
+      return false;
+    }
+    const quotaStausChanged = accessKey.isOverQuota;
+    accessKey.quota = undefined;
+    accessKey.isOverQuota = false;
+    try {
+      this.saveAccessKeys();
+      if (quotaStausChanged) {
+        await this.updateServer();
+      }
     } catch (error) {
       return false;
     }
     return true;
   }
 
   getMetricsId(id: AccessKeyId): AccessKeyMetricsId|undefined {
-    const accessKeyJson = this.getAccessKey(id);
-    return accessKeyJson ? accessKeyJson.metricsId : undefined;
+    const accessKey = this.getAccessKey(id);
+    return accessKey ? accessKey.metricsId : undefined;
+  }
+
+  // Compares access key usage with collected metrics, marking them as under or over quota.
+  async enforceAccessKeyQuotas() {
+    let quotaStatusChanged = false;
+    for (const accessKey of this.accessKeys) {
+      quotaStatusChanged = quotaStatusChanged || await this.updateAccessKeyQuotaStaus(accessKey);
+    }
+    if (quotaStatusChanged) {
+      this.updateServer();
+    }
+  }
+
+  // Updates `accessKey` quota status by comparing its usage with collected metrics. Returns whether
+  // the quota status changed.
+  private async updateAccessKeyQuotaStaus(accessKey: AccessKey): Promise<boolean> {
+    if (!accessKey.quota) {
+      return false;  // Don't query the usage of access keys without quota.
+    }
+    const usage =
+        await this.metrics.getOutboundByteTransfer(accessKey.id, accessKey.quota.windowHours);
+    const usageBytes = usage.bytesTransferredByUserId[accessKey.id];
+    const isOverQuota = usageBytes > accessKey.quota.quotaBytes;
+    logging.debug(`Enforcing quota for access key ${accessKey.id}. Quota: ${
+        JSON.stringify(accessKey.quota)}, usage: ${usageBytes}B, isOverQuota: ${isOverQuota}`);
+    const quotaStatusChanged = isOverQuota !== accessKey.isOverQuota;
+    accessKey.isOverQuota = isOverQuota;
+    return quotaStatusChanged;
   }
 
   private updateServer(): Promise<void> {
-    return this.shadowsocksServer.update(this.keyConfig.data().accessKeys.map((e) => {
-      return {id: e.id, port: e.port, cipher: e.encryptionMethod, secret: e.password};
-    }));
+    const serverAccessKeys = this.accessKeys.filter(e => !e.isOverQuota).map(e => {
```

Comment:
It might be nice to take this opportunity to use "key" or something more descriptive than "e"

Reply:
Done in 2105d3f.